src/antelope/wallets/ual/MetakeepUAL.ts

```diff
@@ -0,0 +1,534 @@
+        this.accountSelector = accountSelector;
+    }
+
+    setAccountNameSelector(accountNameSelector: MetakeepNameAccountSelector) {
+        this.accountNameSelector = accountNameSelector;
+    }
+
+    saveCache() {
+        metakeepCache.saveCache();
+    }
+
+    async init() {
+        //
+    }
+
+    setUserCredentials(credentials: UserCredentials): void {
+        this.userCredentials = credentials;
+        metakeepCache.setLogged(credentials.email);
+    }
+
+    /**
+     * Resets the authenticator to its initial, default state then calls init method
+     */
+    reset() {
+        this.init();
+    }
+
+    /**
+     * Returns true if the authenticator has errored while initializing.
+     */
+    isErrored() {
+        return false;
+    }
+
+    getName() {
+        return metakeep_name;
+    }
+
+    /**
+     * Returns a URL where the user can download and install the underlying authenticator
+     * if it is not found by the UAL Authenticator.
+     */
+    getOnboardingLink() {
+        return '';
+    }
+
+    /**
+     * Returns error (if available) if the authenticator has errored while initializing.
+     */
+    getError(): UALError | null {
+        return null;
+    }
+
+    /**
+     * Returns true if the authenticator is loading while initializing its internal state.
+     */
+    isLoading() {
+        return this.loading;
+    }
+
+    /**
+     * Returns the style of the Button that will be rendered.
+     */
+    getStyle() {
+        return {
+            // An icon displayed to app users when selecting their authentication method
+            icon: 'no-icon',
+            // Name displayed to app users
+            text: metakeep_name,
+            // Background color displayed to app users who select your authenticator
+            background: '#030238',
+            // Color of text used on top the `backgound` property above
+            textColor: '#FFFFFF',
+        };
+    }
+
+    /**
+     * Returns whether or not the button should render based on the operating environment and other factors.
+     * ie. If your Authenticator App does not support mobile, it returns false when running in a mobile browser.
+     */
+    shouldRender() {
+        return true;
+    }
+
+    /**
+     * Returns whether or not the dapp should attempt to auto login with the Authenticator app.
+     * Auto login will only occur when there is only one Authenticator that returns shouldRender() true and
+     * shouldAutoLogin() true.
+     */
+    shouldAutoLogin() {
+        return true;
+    }
+
+    /**
+     * Returns whether or not the button should show an account name input field.
+     * This is for Authenticators that do not have a concept of account names.
+     */
+    async shouldRequestAccountName() {
+        return false;
+    }
+
+    async createAccount(publicKey: string): Promise<string> {
+        const suggestedName = await this.accountNameSelector.selectAccountName();
```

Comment:
```ts
const suggestedName = await this.accountNameSelector.selectAccountName();
```

That line is a callback to the external website to stop the process, allowing the user to select an unused account name for the new account creation. This is optional, and if not overridden, a default empty string will be returned without interaction, leading to the endpoint being called with no suggested account name, resulting in the creation of a completely random account.